### PR TITLE
Reusing logic with custom hooks 2

### PR DESCRIPTION
**Pull request by imnotmoon**
asdf

### CODE REVIEW

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 1]
```diff
@@ -0,0 +1,363 @@
+import {Code} from "../../../src/components/Code";
+
+:::info Deep Dive
+# Custom hook이 높은 수준의 용례들에 대해 집중하도록 하세요
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]:
```suggestion
# Custom hook을 높은 수준에서 사용하세요
```

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+여러분의 custom hook의 이름을 정하는것부터 시작해봅시다.
+만약 명확한 이름을 고르는데 어려움을 느낀다면, effect가 컴포넌트의 로직과 너무 깊은 관계를 맺고 있기 때문일 수도 있습니다.
+그리고 아직 추출될 준비가 되지 않은 것이죠.
+
+이상적으로 여러분의 custom hook의 이름은 코드를 한 줄도 쓸줄 모르는 사람이 보더라도 custom hook이 무엇을 하고 무엇을받고,
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]:
```suggestion
이상적으로 여러분의 custom hook의 이름은 코드를 한 줄도 쓸줄 모르는 사람이 보더라도 custom hook이 무엇을 하고 무엇을 받고,
```

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+외부의 시스템과 싱크를 맞추려 할 때 여러분의 custom hook의 이름은 아마 조금 더 기술적이거나 그 시스템에 대한 전문용어가 들어갈 수 있습니다.
+그 시스템에 친숙한 사람이 봤을때는 충분히 좋은 이름입니다.
+
+- ✅ useMediaQuery(query)
+- ✅ useSocket(url)
+- ✅ useIntersectionObserver(ref, options)
+
+**Custom hook이 높은 수준의 용례들에 집중할 수 있게 해주세요.** Custom hook이 단순히 `useEffect` API를 감싸거나 이의 대안 정도밖에 안되는
+"생명주기" hook이 되어서는 안됩니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]:
```suggestion
** Custom hook을 높은 수준에서 사용하세요.** Custom hook이 단순히 `useEffect` API를 편의를 위해 감싸거나 이의 대안 정도밖에 안되는 "생명주기" hook을 만들거나 사용하는 것을 피해야 합니다.
```

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+
+- 🔴 useMount(fn)
+- 🔴 useEffectOnce(fn)
+- 🔴 useUpdateEffect(fn)
+
+예를 들어, 이 `useMount` hook은 어떤 코드가 "마운트 시점에" 돌아갈 수 있도록 해줍니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]:
```suggestion
예를 들어, 이 `useMount` hook은 코드가 "마운트 시점에" 돌아갈 수 있도록 해줍니다.
```

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    fn();
+  }, []); // 🔴 React Hook useEffect has a missing dependency: 'fn'
+}
+```
+
+**`useMount`같은 커스텀한 "생명주기" hook은 React의 패러다임에 맞아 떨어지지 않습니다.**
+예를 들어, 이 예시 코드는 잘못된 부분이 있습니다 (`roomId`나 `severUrl`의 변화에 대해 "반응"하지 않습니다).
+하지만 linter는 반응하지 않습니다. linter는 `useEffect`의 직접적인 호출에 대해서만 검사하기 때문입니다.
+linter는 여러분의 hook에 대해 알지 못합니다.
+
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx:596]: "If you’re writing an Effect, start by using the React API directly:" 빠졌습니다!

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+  useEffect(() => {
+    post('/analytics/event', { eventName: 'visit_chat', roomId });
+  }, [roomId]);
+
+  // ...
+}
+```
+
+이어서, 서로 다른 고수준 용례에 대해 custom hook을 추출할 수 있습니다(안해도 괜찮습니다).
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx:619]:
```suggestion
이어서, 서로 다른 고수준 사용처에 대해 custom hook을 추출할 수 있습니다(안해도 괜찮습니다).
```

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+  // ✅ custom hook들이 목적을 드러내는 이름을 갖고 있기 때문에 훌륭한 코드입니다.
+  useChatRoom({ serverUrl, roomId });
+  useImpressionLog('visit_chat', { roomId });
+  // ...
+}
+```
+
+**좋은 custom hook은 코드의 호출을 선언적으로 만들고 스스로 무엇을 하는지 명확하게 알고 있습니다.**
+예를 들어, `useChatRoom(options)`는 오직 채팅방과 연결만 할 수 있고, `useImpressionLog(eventName, extraData)`는 오직 분석 툴에 로그를 보내는 일만 할 수 있습니다.
+만약 여러분의 custom hook API가 용례를 제한하지 않고 매우 추상적이라면, 장기적으로는 풀기 어려운 문제들을 많이 만들 것입니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]:
```suggestion
만약 여러분의 custom hook API가 사용되는 상황을 제한되지 않고 매우 추상적이라면, 장기적으로는 풀기 어려운 문제들을 많이 만들 것입니다.
```

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+용례에 대한 충분히 훌륭한 내장 솔루션이 없을때 "React 외부와 연결"하기 위해 effect를 이용합니다.
+한때 React 팀의 목표는 더 구체적인 문제들에 대해 더 구체적인 해결방법을 제시함으로써 여러분의 app 내에서 effect의 사용을 최대한 줄이는 것이었습니다.
+Effect를 custom hook으로 감싸는 것은 이런 해결 방법들이 유효할때 코드를 쉽게 업그레이드할 수 있도록 도왔습니다.
+
+아래 예시로 돌아가봅시다.
+
+<Code src="https://codesandbox.io/embed/peaceful-leftpad-9x38on?fontsize=14&hidenavigation=1&theme=dark" />
+
+위 예시에서 `useOnlineStatus`는 한 쌍의 [`useState`](https://react.dev/reference/react/useState)와 [`useEffect`](https://react.dev/reference/react/useEffect)로 작성되어 있습니다.
+하지만 이건 최선의 해결방법이 아닙니다. 이 hook이 고려하지 않고 있는 몇개의 edge case들이 있스빈다.
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]:
```suggestion
하지만 이건 최선의 해결방법이 아닙니다. 이 hook이 고려하지 않고 있는 몇개의 edge case들이 있습니다.
```

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+순수한 effect를 통해 모든 컴포넌트에 각각 작성된 코드보다 훨씬 적은 수정만으로 우리가 원하는 방향으로 마이그레이션 할 수 있을 것입니다.
+하지만 오래된 접근법도 여전히 잘 동작합니다. 순수한 effect를 적는 코드가 충분히 괜찮다면 굳이 바꿀 필요는 없습니다.
+
+:::
+
+### 한가지 방법만 존재하지 않습니다.
+
+브라우저의 [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) API를 이용해
+fade-in 애니메이션을 *처음부터* 만든다고 가정해봅시다.
+아마 애니메이션 루프를 설정하는 effect부터 만들 것입니다. 각 애니메이션의 프레임에서는 [ref에 할당된] DOM 노드의 opacity를 `1`까지 변경할 것입니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]: 링크가 빠졌습니다!

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+
+<Code src="https://codesandbox.io/embed/sad-wozniak-iwzkkv?fontsize=14&hidenavigation=1&theme=dark" />
+
+`useFadeIn` 코드를 유지해도 괜찮지만, 조금 더 리팩토링 하길 원할 수도 있습니다.
+예를 들어, `useFadeIn`에서 애니메이션 루프를 세팅하는 로직을 꺼내 `useAnimationLoop` custom hook으로 옮길 수도 있습니다.
+
+<Code src="https://codesandbox.io/embed/condescending-merkle-w6nwnb?fontsize=14&hidenavigation=1&theme=dark" />
+
+마찬가지로 이렇게 *할 필요는* 없습니다. 일반 함수를 작성하듯, 궁극적으로는 코드의 서로 다른 부분에 경계선을 어떻게 그릴지를 결정하게 됩니다.
+혹은 전혀 다른 접근을 할 수도 있스빈다. Effect 안에 로직을 두지 않고 로직의 가장 명령적인 부분들을 JavaScript [class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)의 내부로 옮길 수도 있습니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]:
```suggestion
혹은 전혀 다른 접근을 할 수도 있습니다. Effect 안에 로직을 두지 않고 로직의 가장 명령적인 부분들을 JavaScript [class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)의 내부로 옮길 수도 있습니다.
```

**Review by hyehyeonmoon — COMMENTED** [review]
리뷰 끝!

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx (hunk 1 above)
Comment (imnotmoon) [reply docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]: 훨씬 자연스럽네요 👍

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx (hunk 6 above)
Comment (imnotmoon) [reply docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx:619]: use case와 사용처는 다른 표현인것 같습니다.
고수준 사용처는 말이 안돼요.

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx (hunk 9 above)
Comment (imnotmoon) [reply docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx]: [09c4d3a](https://github.com/LINE-NEXT-LEVEL/client/pull/46/commits/09c4d3a2489fbd31afb582ecdd092e8a873c1c96)

추가했습니다!

File: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx (hunk 5 above)
Comment (imnotmoon) [reply docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx:596]: [412a8bc](https://github.com/LINE-NEXT-LEVEL/client/pull/46/commits/412a8bcfb7f7913368730b59575fbf2738b010ea)

추가했습니다 :)

**Review by hyehyeonmoon — APPROVED** [review]
👍👍👍

**Pull request merged**